### PR TITLE
First pass at input gaurdrails

### PR DESCRIPTION
**Pull request by sidravi1**
Reviewer: @suzinyou 

Estimate: 1 hr

----

# First pass at input gaurdrails

## Ticket

Fixes: [AAQ-223](https://idinsight.atlassian.net/browse/AAQ-223)

## Description, Motivation and Context

We need to add guardrails for safety.
See [this document](https://idinsight.atlassian.net/wiki/spaces/PD/pages/2332360823/End-to-end+LLM+Question-Answer+flow) for high-level approach.

### Changes

Added a few additional steps for llm-response.
Switched to gpt-4. It doesn't work so well with gpt-3-turbo - the model just doesn't follow instructions as well.
I also tried to run it with marvin `ai_classifier` but it didn't work so well. I'll try it again in a future PR.

### Why

Safety and better performance.
Also, first pass at language support.

## How Has This Been Tested?

Existing test cases pass. 
In the next PR, I will be writing test cases specifically for testing the new functions/prompts to make sure it is indeed
identifying language, translating, doing safety check correctly.

## Next steps

Within this sprint:
* Add test cases
* Add documentation

Next sprint:
* Add CoT and other ideas from the design doc

## Checklist:

This checklist is a useful reminder of small things that can easily be forgotten.
Put an `x` in all the items that apply and remove any items that are not relevant to this PR.

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests (if applicable)
- [x] I have written [good commit messages][1]
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: core_backend/app/configs/llm_prompts.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        (b) inappropriate language - racist, sexist, offensive, or insulting language.
+        Assess the text above for the presence of these two.
+        Respond strictly with {" or ".join(SafetyClassification._member_names_)} only.
+        Answer should be a single word.
+"""
+
+# ----  Language identification bot
+
+
+class IdentifiedLanguage(Enum):
```
Comment (sidravi1) [inline core_backend/app/configs/llm_prompts.py:24]: By using Enum, we are catching cases where LLM doesn't follow instructions and returns something outside of the provided options.

Comment (sidravi1) [reply core_backend/app/configs/llm_prompts.py:24]: You can just update languages here to add support for other languages.... in theory.

File: core_backend/app/routers/question_answer.py [hunk 2]
```diff
@@ -38,45 +46,50 @@ async def llm_response(
...
     response = UserQueryResponse(
         query_id=user_query_db.query_id,
-        content_response=content_response,
-        llm_response=llm_text_response,
+        content_response=None,
+        llm_response="Sorry, we cannot do that.",
         feedback_secret_key=feedback_secret_key,
     )
+
+    # -- pre-process input
```
Comment (sidravi1) [inline core_backend/app/routers/question_answer.py:64]: Will refactor in future PR to take each of these out to a separate function. Make it a little easier to understand

File: core_backend/app/routers/question_answer.py [hunk 3]
```diff
@@ -38,45 +46,50 @@ async def llm_response(
...
+    if not input_is_safe(user_query_refined):
+        return response
+
+    user_query_refined = identify_language(user_query_refined)
+    if user_query_refined.original_language == IdentifiedLanguage.UNKNOWN:
+        response.llm_response = (
+            "Sorry, we are unable to understand your question. "
```
Comment (sidravi1) [inline core_backend/app/routers/question_answer.py:71]: Will also move these out to a different config in a future PR. Would be nice to allow responses to be customizable without messing with code

File: core_backend/app/routers/question_answer.py [hunk 4]
```diff
@@ -93,25 +106,36 @@ async def embeddings_search(
...
-    # add to query database
     user_query_db = await save_user_query_to_db(
         asession, feedback_secret_key, user_query
     )
 
-    # get FAQs from vector db
+    user_query_refined = UserQueryRefined(
+        **user_query.model_dump(), query_text_original=user_query.query_text
+    )
+    user_query_refined = identify_language(user_query_refined)
```
Comment (sidravi1) [inline core_backend/app/routers/question_answer.py:116]: If you can identify it as a language that is not English, translate it.

File: core_backend/app/llm_call/llm_rag.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from ..configs.llm_prompts import ANSWER_QUESTION_PROMPT
+from .utils import _ask_llm
+
+
+def get_llm_rag_answer(question: str, faq_data: str) -> str:
```
Comment (sidravi1) [inline core_backend/app/llm_call/llm_rag.py:5]: Note that the answer is at the moment in English. It does not ask it to generate it in the language identified. We could do that though!

File: core_backend/app/llm_call/parse_input.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def identify_language(question: UserQueryRefined) -> UserQueryRefined:
+    """
+    Identifies the language of the question.
+    """
+
+    question.original_language = getattr(
+        IdentifiedLanguage, _ask_llm(question.query_text, IDENTIFY_LANG_INPUT)
+    )
```
Comment (suzinyou) [inline core_backend/app/llm_call/parse_input.py:39]: Would be nice to somehow associate `IdentifiedLanguage` with `IDENTIFY_LANG_INPUT`, and `SafetyClassification` with `CHECK_INPUT_SAFETY`and abstract out the classification. I guess that's what marvin's AI Classifier does but apparently poorly :') So I guess (in sketch)

```
class Classifier:
  prompt: str
  classes: Enum
  def classify(question):
    return getattr(self.classes, _ask_llm(question.query_text, self.prompt)
```

File: core_backend/app/schemas.py [hunk 7]
```diff
@@ -14,6 +16,17 @@ class UserQueryBase(BaseModel):
     query_text: str
     query_metadata: dict = {}
 
+    model_config = ConfigDict(from_attributes=True)
```
Comment (suzinyou) [inline core_backend/app/schemas.py:19]: What does this do? Could not understand the [docs](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.from_attributes) :/

File: core_backend/app/routers/question_answer.py (hunk 3 above)
Comment (suzinyou) [reply core_backend/app/routers/question_answer.py:71]: Yes I think it makes sense to bake prompts into code but response formats would ideally be customized by content managers thru admin app

**Review by suzinyou — COMMENTED** [review]
Thanks Sid! `gaurdrails` :D
Leaving initial comments
Also what is the point of the package guardrails?

**Review by suzinyou — APPROVED** [review]
All tests ran smoothly! Thanks for taking the first stab at safety features!!!

File: core_backend/app/llm_call/parse_input.py (hunk 6 above)
Comment (sidravi1) [reply core_backend/app/llm_call/parse_input.py:39]: Oh interesting idea. Let me see if I can do this quickly in this PR else in a future one

File: core_backend/app/schemas.py (hunk 7 above)
Comment (sidravi1) [reply core_backend/app/schemas.py:19]: It allows you to access the properties with either dot notation or like a dict.

**Pull request merged**